dotCMS/src/main/java/com/dotcms/publisher/util/DependencyManager.java

```diff
@@ -1490,4 +1341,176 @@ Set getContainers() {
+			if (!started) {
+				return;
+			}
+
+			if (assetsAlreadyProcessed != null && !this.alreadyProcess(assetKey, assetType)) {
+				Logger.debug(DependencyProcessor.class, () -> String.format("%s: Putting %s in %s",
+						Thread.currentThread().getName(), assetKey, assetType));
+
+				queue.add(new DependencyProcessorItem(assetKey, assetType));
+				addSet(assetKey, assetType);
+			}
+		}
+
+		private synchronized void addSet(final String assetKey, final AssetTypes assetTypes) {
+			Set<String> set = assetsAlreadyProcessed.get(assetTypes);
+
+			if (set == null) {
+				set = new HashSet<>();
+				assetsAlreadyProcessed.put(assetTypes, set);
+			}
+
+			set.add(assetKey);
+		}
+
+		void start(){
+			Logger.debug(DependencyProcessor.class, "starting");
+			queue = new LinkedBlockingDeque();
+			assetsAlreadyProcessed = new HashMap<>();
+
+			int poolSize = Config.getIntProperty("NUMBER_THREAD_TO_EXECUTE_BUNDLER", 10);
+
+			threads = new ArrayList<>();
+
+			for (int i = 0; i < poolSize; i++) {
+				final DependencyThread thread = new DependencyThread();
+				threads.add(thread);
+
+				Logger.debug(DependencyProcessor.class, () -> String.format("Starting %s", thread.getName()));
+				thread.start();
+			}
+
+			started = true;
+		}
+
+		public void waitUntilFinish(){
+			Logger.debug(DependencyProcessor.class, () -> String.format("%s: Wait until all finish",
+					Thread.currentThread().getName()));
+
+			while(!isFinish()) {
+				Logger.debug(DependencyProcessor.class, () -> String.format("%s: it is not finish yet",
+						Thread.currentThread().getName()));
+				waitForDependencies();
+			}
+
+			Logger.debug(DependencyProcessor.class, () -> String.format("%s: it is finish",
+					Thread.currentThread().getName()));
+			this.stop();
+		}
+
+		private synchronized void waitForDependencies() {
+			try {
+				wait(TimeUnit.MINUTES.toMillis(1));
+			} catch (InterruptedException e) {
+				Logger.error(DependencyProcessor.class, e.getMessage());
+			}
+		}
+
+		private synchronized void notifyMainThread() {
+			notify();
+		}
+
+		private boolean isFinish(){
+			return queue == null || (queue.isEmpty() && allThreadWaiting());
+		}
+
+		private boolean allThreadWaiting() {
+			for (final DependencyThread thread : threads) {
+				if (!thread.getWaitingState()) {
+					return false;
+				}
+			}
+
+			return true;
+		}
+
+		void stop(){
+			this.finished = true;
+
+			if (threads != null) {
+				for (final Thread thread : threads) {
+					thread.interrupt();
+				}
+			}
+		}
+
+		public boolean alreadyProcess(final String assetKey, final AssetTypes assetTypes) {
+			final Set<String> set = assetsAlreadyProcessed.get(assetTypes);
+			return set != null && set.contains(assetKey);
+		}
+
+		private class DependencyThread extends Thread {
+			boolean waitingForSomethingToProcess = false;
+
+			@Override
+			public void run() {
+				while (true) {
+					try {
+						DependencyProcessorItem dependencyProcessorItem = queue.poll();
```

Comment:
How this finish? It looks if will be eternal until interrupted

Reply:
done https://github.com/dotCMS/core/pull/19955/commits/803a1c5b4aae8f0a873736babf79c2ba79b734b3